discard changes

Byron opened the pull request:
Add V3 facilities for discarding changes in the worktree or index.

### Tasks

* [x] reset whole files (with safety-check as change must be matched)
    - [x] addition
    - [x] deletion
        - [x] symlinks
        - [x] submodules
        - [x] executable bit
    - [x] modification
        - [x] symlinks
        - [x] submodules
        - [x] executable bit
        - [x] type-change
    - [x] fix: re-generate submodule from local clone only
    - [x] restore type change from file to dir (dir is to be restored)
    - [x] replace file with directory (with tests for single and multi-discard)
    - [x] renamed file
        - [x] onto nothing
        - [x] onto directory
        - [x] onto other file
    - [x] renamed directory onto file
    - [x] conflicted (aren't listed, so don't need handling, double-check)
    - [x] walk-through-symlink-safety
* [x] index-update support (`.git/index` needs to represent the worktree) as needed
* [x] update `gix` to point to `main`
* [x] CLI integration
* [x] fix disk-space issue by avoiding an additional docker invocation, and fix the now failing tests.
* [x] `tauri`-integration to satisfy current UI capabilities
* [x] ~~restore rust-test parallelism~~ - no need, the cache works well and a full test-run is now below 5 minutes

### Next PR

* [ ] reset changed hunks in a file
* [ ] reset individual lines


### Notes for the Reviewer

* I flattened the Rust test setup, removing indirections and seemingly speeding it up, while reducing the chance that the runner runs out of disk space. For this PR, this now happened all the time, probably because running through docker consumes more disk space? Having a 'dumber' CI setup also seems like a win to me, despite less elegant handling of dependencies.
    - Installing dependencies takes 34s, and pulling the container takes 36. However, the 36s contains the Rust cache, which isn't included in the 34s count.
    - Overall Usage for Rust jobs is reduced from 36 minutes to ~~24~~ 19 minutes, and as less jobs are run more PRs can run in parallel.
    - Admittedly, removing docker from the mix was just an attempt to make CI work again, and it does indeed work now.
    - An uncached Rust test CI run can take 12 minutes, but with a cache this goes down to 4m 50s, which is less than the usual docker run (which seems a little slower under the same circumstances) <img width="285" alt="Screenshot 2025-03-18 at 14 50 24" src="https://github.com/user-attachments/assets/c8d83995-313b-4be7-80e4-c26f12d6ea5d" />
    - A CI run still takes just a little more than 5 minutes, so it remains at the sweet spot.

* It turns out that the frequent CI failures were caused by disguised out-of-disk-space errors. They appear to be an LLVM-LD crash, but apparently that is just a C-Programs way of saying that [the disk is full](https://github.com/llvm/llvm-project/issues/98988#issuecomment-2235270931).
    - Clearling the GH caches usually does the trick, but it's bound to re-occur.

### Shortcomings

* Index-handling is low-level and I think there needs to be something like a tree-editor, but for indices. Maybe it's a mix of using the pretty-neat tree-editor, and a way to transfer index information from one index to another, similar to 'apply', to avoid loosing information.
* UX: submodules aren't automatically checked out if they are part of a directory. It's a relatively rare case as well.
* UX: Around renames, there can be a special case that makes the user discard a set of files twice before they go away. It's rare, and it's fixable if one wants to put in the time.

### Out of Scope

* snapshot integration - should be left to @krlvi to warm up with the new code.

### Research

* **V2 implementation**
    - spread across `unapply_lines`, `unapply_ownership` and `reset_files`

Byron committed:
update `gix` to the latest version

krlvi commented:
Awesome! One random thought - when it comes time to wire this as a tauri /CLI command, we could consider having the guard passed instead of it being created deeper in the call stack. Mainly because the snapshotting would need the same guard too - here's what i have in mind https://github.com/gitbutlerapp/gitbutler/commit/d3e1f42cc7160320c7cde35ed99c3f57fa528b64

Byron commented:
Right, thanks for the reminder! I remember the dead-lock problem that taught us that lesson 😅.
Maybe it's fair to say that these plumbing crates should never create this guard themselves, so either CLI or Tauri have to create it.

Byron committed:
Add V3 support for discarding files

For now, no hunks or hunk-ranges.

Byron committed:
Make the new `discard` available through `tauri`

Byron committed:
Add `discard-changes` to but-cli

That way the discard implementation can be triggered in the real world.

Byron committed:
attempt to speedup CI runs by not building criterion

It seems to take quite some time on CI even though it's only for benches,
which already have a feature toggle.

Also remove unused dependencies everywhere.

Byron committed:
Avoid using separate docker invocation for simplicity, speed and disk-space.

We run out of disk-space so often now that the system is barely usable.
To keep the change minimal, this is only applied to Rust checks.

Byron committed:
Fix CI failures that happen due to docker-avoidance.

It's an interesting side-effect that running in Docker seems to work
better, maybe due to running as root?

Try to emulate what seems to be happening on CI by changing the desired
umask, which usually does the right thing when running locally.

krlvi commented:
Looking great, thanks Byron!
